Fixes format inconsistencies with docker for certain history fields

rbagd opened the pull request:
Closes #17767
Closes #17768

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Fixes a format inconsistency where podman and docker return slightly different output for `history --format {{.CreatedAt}}` and `history --format {{.Size}}` (#17767, #17768)
```

rbagd commented:
CI indicates that one of the `bats` tests failed, namely `podman image history Created`. Besides a regex check which is fixable, it also requires that `podman image list --format {{.CreatedAt}}` would be the same as `podman image history --format {{.CreatedAt}}`. But Docker itself outputs them in a different format, so it's unclear now if this comparison should be dropped to favor docker compatibility... :man_shrugging: 

Assuming PR is merged,

```
$ docker image list --format {{.CreatedAt}} busybox | head -n 1
2023-03-17 02:35:43 +0100 CET
$ docker image history --format {{.CreatedAt}} busybox | head -n 1
2023-03-17T02:35:43+01:00

$ podman image list --format {{.CreatedAt}} busybox | head -n 1
2023-03-17 01:35:43 +0000 UTC
$ podman image history --format {{.CreatedAt}} busybox | head -n 1
2023-03-17T02:35:43+01:00

```

rhatdan commented:
I am fine with fixing this, but I will allow @vrothberg and @mheon to chime in.

mheon commented:
I'd prefer to retain the test - it seems like it could be changed to parse the fields back to a `time.Time` and compare them that way. LGTM otherwise.

vrothberg reviewed (commented):
The changes look good to me.

One system test isn't passing (see https://cirrus-ci.com/task/5421900956958720?logs=main#L284).  @edsantiago, we probably need to your scripting skills.  I tried normalizing the two dates but one fails with

`$ date -d '2022-10-18 16:28:08 +0000 UTC'`

Luap99 commented on test/e2e/history_test.go:
```diff
@@ -44,6 +44,13 @@ var _ = Describe("Podman history", func() {
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(Exit(0))
+		lines := session.OutputToStringArray()
+		Expect(lines).ToNot(BeEmpty())
+		Expect(lines[0]).ToNot(ContainSubstring(" "))
```
Please do not make negative checks. You should check for the real output you expect. I know there a bunch of bad examples in the existing tests but we should do a better job for new tests.

If the numbers are not consistent you can use `To(MatchRegexp(...))` to match a regex instead.
Also just call `session.OutputToString()` to get the output as one string.

rbagd replied:
No problem, I update the PR with an extra regex.

edsantiago commented:
The failing system test is: `CreatedSince and CreatedAt from image history should == image list`. TL;DR this PR makes `podman image list` and `podman image history` inconsistent with each other:
```console
$ for i in list history;do echo;bin/podman image $i --format '{{.CreatedAt}}' quay.io/libpod/testimage:20221018;done

2022-10-18 16:28:08 +0000 UTC     <--- until now, this has been the format for both

2022-10-18T10:28:08-06:00         <--- new, changed format
2022-10-18T10:28:08-06:00
2022-10-18T10:28:08-06:00
2022-10-18T10:28:08-06:00
2022-10-18T10:28:20-06:00
2022-10-18T10:28:20-06:00
2022-10-18T10:28:20-06:00
```

eriksjolund commented:
In the commit message it's written

```
Closes #17767 and #17768 
```

The GitHub linking only works for the first one (17767) but not the second one (17768)

I would guess the commit message needs to be written like this for it to work

```
Closes #13797 
Closes #17768 
```

Quote:

```
Multiple issues	Use full syntax for each issue
```
from 
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

rbagd commented:
I updated the relevant system test to maintain the equality check of `podman image history` and `podman image list` output for `CreatedAt` and `CreatedSince`. Please let me know if this look good enough. To make the whole thing more readable I switched field separator in the test to `;` from `--`.

CI failed for one of the jobs but it's a [runtime error which does not seem related to the PR](https://cirrus-ci.com/task/5606446105100288?logs=main#L454).

Luap99 commented:
@edsantiago PTAL at system tests

edsantiago commented on test/system/110-history.bats:
```diff
@@ -57,15 +57,26 @@ size      | -\\\?[0-9]\\\+
+    imagehist_at=$(date --rfc-3339=seconds -f <(echo "$from_imagehistory" | cut -d';' -f2))
+
+    assert "$imagehist_at" == "$imagelist_at" \
+           "CreatedAt from image history should == image list"
```
I find all the `cut/sed`s a little hard to read, and they're actually unnecessary. I also think it's important to explain what's going on, because this is now super-confusing. Suggestion:
```bash
@test "podman image history Created" {
    # Values from image LIST
    run_podman image list --format '{{.CreatedSince}}\n{{.CreatedAt}}' $IMAGE
    imagelist_since="${lines[0]}"
    imagelist_at="${lines[1]}"
    assert "${imagelist_since}" =~ "^[0-9].* ago" \
           "image list: CreatedSince looks reasonable"
    assert "${imagelist_at}" =~ "^[0-9]+-[0-9]+-[0-9]+ [0-9:]+ \+0000 UTC\$" \
           "image list: CreatedAt looks reasonable"

    # Values from image HISTORY. For docker compatibility, this command now
    # honors $TZ (#18213) for CreatedAt.
    TZ=UTC run_podman image history --format '{{.CreatedSince}}\n{{.CreatedAt}}' $IMAGE
    imagehistory_since="${lines[0]}"
    imagehistory_at="${lines[1]}"

    assert "$imagehistory_since" == "$imagelist_since" \
           "CreatedSince from image history should == image list"

    # More docker compatibility: both commands emit ISO8601-ish dates but
    # with different separators so we need to compare date & time separately.
    assert "${imagehistory_at:0:10}" == "${imagelist_at:0:10}" \
           "CreatedAt (date) from image history should == image list"
    assert "${imagehistory_at:11:8}" == "${imagelist_at:11:8}" \
           "CreatedAt (time) from image history should == image list"
}
```

rbagd replied:
This looks very nice! I'll push an update later today. Thanks for looking into it!

edsantiago commented:
LGTM. Thanks @rbagd.

@containers/podman-maintainers PTAL, especially the docker compatibility (I don't have a good way to test docker).

eriksjolund commented:
@rbagd Thanks for adjusting the commit message so that it now says

```
Closes #17767
Closes #17768
```

Unfortunately it didn't fix the the GitHub linking. Sorry for my misleading information. 
I made a test in another GitHub repository and saw that we can fix the GitHub linking by editing the [pull request's description](https://github.com/containers/podman/pull/18213#issue-1669357754) and replace

```
Closes #17767 and #17768
```

with 

```
Closes #17767
Closes #17768
```

rbagd commented:
It seems to be working! Both issues appear on the right sidebar now as expected.

TomSweeneyRedHat commented:
LGTM

rhatdan commented:
/approve
/lgtm
